src/transport/assembly_convection.hh

```diff
@@ -0,0 +1,495 @@
+    : AssemblyBase<dim>(0), eq_fields_(eq_fields), eq_data_(eq_data) {
+        this->active_integrals_ = ActiveIntegrals::edge | ActiveIntegrals::coupling;
+        this->used_fields_ += eq_fields_->flow_flux;
+    }
+
+    /// Destructor.
+    ~MatrixMpiAssemblyConvection() {}
+
+    /// Initialize auxiliary vectors and other data members
+    void initialize(ElementCacheMap *element_cache_map) {
+        this->element_cache_map_ = element_cache_map;
+
+        fe_ = std::make_shared< FE_P_disc<dim> >(0);
+        UpdateFlags u = update_values | update_gradients | update_side_JxW_values | update_normal_vectors | update_quadrature_points;
+        fe_values_side_.initialize(*this->quad_low_, *fe_, u);
+        fe_values_vec_.resize(eq_data_->max_edg_sides);
+        for (unsigned int sid=0; sid<eq_data_->max_edg_sides; sid++)
+        {
+            fe_values_vec_[sid].initialize(*this->quad_low_, *fe_, u);
+        }
+
+    }
+
+    /// Assembles the fluxes between sides of elements of the same dimension.
+    inline void edge_integral(RangeConvert<DHEdgeSide, DHCellSide> edge_side_range) {
+        ASSERT_EQ_DBG(edge_side_range.begin()->element().dim(), dim).error("Dimension of element mismatch!");
+
+        int sid=0, s1, s2;
+        edg_flux = 0.0;
+        for( DHCellSide edge_side : edge_side_range )
+        {
+            fe_values_vec_[sid].reinit(edge_side.side());
+            auto p = *( this->edge_points(edge_side).begin() );
+            flux = arma::dot(eq_fields_->flow_flux(p), fe_values_vec_[sid].normal_vector(0)) * fe_values_vec_[sid].JxW(0);
+            if (flux > 0.0) {
+                eq_data_->cfl_flow_[edge_side.cell().local_idx()] -= (flux / edge_side.element().measure() );
+                edg_flux += flux;
+            }
+            ++sid;
+        }
+
+        s1=0;
+        for( DHCellSide edge_side1 : edge_side_range )
+        {
+            s2=-1; // need increment at begin of loop (see conditionally 'continue' directions)
+            auto p1 = *( this->edge_points(edge_side1).begin() );
+            flux = arma::dot(eq_fields_->flow_flux(p1), fe_values_vec_[s1].normal_vector(0)) * fe_values_vec_[s1].JxW(0);
+            for( DHCellSide edge_side2 : edge_side_range )
+            {
+                s2++;
+                if (s2==s1) continue;
+
+                auto p2 = *( this->edge_points(edge_side2).begin() );
+                flux2 = arma::dot(eq_fields_->flow_flux(p2), fe_values_vec_[s2].normal_vector(0)) * fe_values_vec_[s2].JxW(0);
+                new_i = eq_data_->row_4_el[edge_side1.element().idx()];
+                new_j = eq_data_->row_4_el[edge_side2.element().idx()];
+                if ( flux2 > 0.0 && flux <0.0)
+                    aij = -(flux * flux2 / ( edg_flux * edge_side1.element().measure() ) );
+                else aij =0;
+                MatSetValue(eq_data_->tm, new_i, new_j, aij, INSERT_VALUES);
+            }
+            s1++;
```

Comment:
Set aij to a row vector and call MatSetvalues once per s1 iteration.

MatSetValues(tm, {new_i}, all_elem_dofs, row_values, INSERT_VALUES)